### ISSUE TITLE
[Bug]: isVisible() behaviour does not match Documentation, conditional test paths truthy/falsely

### ISSUE THREAD

**Issue by renhyl** [microsoft/playwright#30230]
### Version

1.42.1

### Steps to reproduce

.waitFor() is NOT waiting for element to appear inside if ()
.isVisible() is waiting for element to appear inside if ()

![image](https://github.com/microsoft/playwright/assets/2118334/605ecd80-3da6-4a80-825b-adfc0b170d79)

https://playwright.dev/docs/api/class-locator#locator-wait-for
as per documentation 
.waitFor({ state: 'attached', timeout: 5000 }) I have also attempted to test different variations of state: 'visible' etc.

state - 'attached' - wait for element to be present in DOM.

### Expected behavior

I would expect .waitFor() to work inside if() same as .isVisible()

### Actual behavior

it appears that .waitFor() is not evaluated inside if ()

### Additional context

https://playwright.dev/docs/api/class-locator#locator-is-visible

in general the Playwright's documentation about appearance and disappearance of elements is rather poor and vague at best.

### Environment

```shell
System:
    OS: Linux 5.15 Ubuntu 20.04.6 LTS (Focal Fossa)
    CPU: (32) x64 AMD Ryzen 9 5950X 16-Core Processor
    Memory: 44.44 GB / 50.98 GB
    Container: Yes
  Binaries:
    Node: 20.11.1 - ~/.nvm/versions/node/v20.11.1/bin/node
    Yarn: 1.22.19 - /work/node_modules/.bin/yarn
    npm: 10.2.4 - ~/.nvm/versions/node/v20.11.1/bin/npm
    pnpm: 8.15.4 - ~/.local/share/pnpm/pnpm
    bun: 1.1.0 - ~/.bun/bin/bun
  IDEs:
    VSCode: 1.87.2 - /home/damian/.vscode-server/bin/863d2581ecda6849923a2118d93a088b0745d9d6/bin/remote-cli/code
  Languages:
    Bash: 5.0.17 - /usr/bin/bash
  npmPackages:
    @playwright/test: ^1.36.2 => 1.42.1
```

**Comment by dgozman** [comment]
@renhyl Could you please be more specific and also share a repro we can run locally?

> I would expect .waitFor() to work inside if() same as .isVisible()

I don't really understand this. `waitFor()` and `isVisible()` are two different methods, they do different things.

Also note that `waitFor()` does return a `Promise<void>`, so checking its return value as in the screenshot will always be falsy. When locator does not appear during the timeout, `waitFor()` will throw instead of returning `false`.

**Comment by Romarionijim** [comment]
I have a problem with the waitfor timeout as well, the timeout option inside the visible state does not have any effect, it does not wait for the element to be visible when setting the timeout,  it fails straight away without waiting a second

example:

  public async waitForElementToBeVisible(locator: (string | Locator)) {
    let locatorType = await this.getTypeOfLocator(locator);
    await locatorType.waitFor({ state: 'visible', timeout: 5000 });
  }
}

**Comment by dgozman** [comment]
@Romarionijim We need a repro to be able to act on this issue.

**Comment by renhyl** [comment]
> @renhyl Could you please be more specific and also share a repro we can run locally?
> 
> > I would expect .waitFor() to work inside if() same as .isVisible()
> 
> I don't really understand this. `waitFor()` and `isVisible()` are two different methods, they do different things.
> 
> Also note that `waitFor()` does return a `Promise<void>`, so checking its return value as in the screenshot will always be falsy. When locator does not appear during the timeout, `waitFor()` will throw instead of returning `false`.

@dgozman thank you for a replay on this issue, unfortunately your statement is not very helpful, please allow me to clarify.
`waitFor` as the **name of the method** suggests the method "waits", thus it has to return `Promise` as a matter of fact it does return Promise: https://github.com/microsoft/playwright/blob/8deb44cd37c3676019de93b6855580b2573d8650/packages/playwright-core/src/client/locator.ts#L342

This is also Documented:
ref: https://playwright.dev/docs/api/class-locator#locator-wait-for

![image](https://github.com/microsoft/playwright/assets/2118334/127f7085-7cdf-487e-b43d-782d5c2d8e55)

To further clarify, I never claimed that `.waitFor()` and `.isVisible()` are same methods, so to re-iterate your words, yes these are two different methods, but if you read the documentation they "kind of" do similar thing but perhaps in different way?
Furthermore according to the documentation `.isVisible()` should not wait but in my testing it appears in fact that it does wait.
![image](https://github.com/microsoft/playwright/assets/2118334/8b02e63f-fd00-48e0-97b4-2ccb53ca0712)

Most importantly, documentation discurages use of `.isVisible()` and implies that `.isVisible()` can be flaky and should be used with [expect(locator).toBeVisible()](https://playwright.dev/docs/api/class-locatorassertions#locator-assertions-to-be-visible)

So to summerize, I am reporting this as bug because:
1. `.waitFor()` is not doing what it should do
2. `.isVisible()` appears to be awaiting regardles of `timeout` ref: https://github.com/search?q=repo%3Amicrosoft%2Fplaywright%20isVisible&type=code

Hope this answers the confusion, @dgozman , thanks once again for the input, I really apricate any help on this issue!

**Comment by renhyl** [comment]
@dgozman here is an example for you that I can reproduce

https://try.playwright.tech/?l=playwright-test&s=6712y0p

```
import { test, expect } from '@playwright/test';

test('basic test', async ({ page }) => {
  await page.goto('https://playwright.dev/');
  // await expect(page.locator('.DocSearch-Button-Keys')).toBeVisible() // works!
  // await expect(page.locator('.DocSearch-Button-Keys')).waitFor() // does not work, but it should
  await expect(page.locator('.DocSearch-Button-Keys')).waitFor({ timeout: 10000 }) // does not work, does not await
});
```
paste it in:
https://try.playwright.tech/

**Comment by dgozman** [comment]
@renhyl Looking at your code snippet, there is no `waitFor()` assertion to be used with `expect()`, only a method on the Locator. The following works for me:

```ts
import { test, expect } from '@playwright/test';

test('basic test', async ({ page }) => {
  await page.goto('https://playwright.dev/');
  await page.locator('.DocSearch-Button-Keys').waitFor({ timeout: 10000 });
});
```

**Comment by renhyl** [comment]
> @renhyl Looking at your code snippet, there is no `waitFor()` assertion to be used with `expect()`, only a method on the Locator. The following works for me:
> 
> ```ts
> import { test, expect } from '@playwright/test';
> 
> test('basic test', async ({ page }) => {
>   await page.goto('https://playwright.dev/');
>   await page.locator('.DocSearch-Button-Keys').waitFor({ timeout: 10000 });
> });
> ```


hmm interesting, I guess this might be a clue that is hiding here.
@dgozman please can you clarify, are you stating that the correct way of using `waitFor` is without `expect` assertion?

I did some more testing and I think I know what is happening..

https://try.playwright.tech/?l=playwright-test&s=amm3c95

waitFor does not return any value that is why putting it inside `if(waitFor)` is futile, 
waitFor for Truthy path appears to void (returned value is `undefined`)
waitFor for Falsely path throws error/exception which kills any further execution

I think this is somewhat "problematic", if that promise returned value once it resolved if it returned Boolean then we could

```
await page.locator('.DocSearch-Button-Keys').waitFor({ state: 'attached', timeout: 500 }).then((val) => {
if (val) {
// truthy
} else {
// falsely 
}
})
```

and this appears to be the case for https://playwright.dev/docs/api/class-locator#locator-is-visible

![image](https://github.com/microsoft/playwright/assets/2118334/d142f4d2-f66b-40e7-81b3-196d2f105d02)

So in order to have conditional path of assertions we could either go with setTimeout on promise callback or wrap that in try/catch..

In summary, it appears waitFor is working correctly (as per documentation), it is `.isVisible()` is not working as per documentation.

**Comment by dgozman** [comment]
@renhyl I have to admit, I am now lost in this conversation. Given your new insights above, could you please step back and reformulate what exactly is the issue you are facing?

**Comment by renhyl** [comment]
@dgozman sure, sorry for all the confusion, I will try my best to clarify. After all the testing I did I have arrived at the conclusion that `isVisible()` is not working correctly because it awaits timeout but it should not according to documentation: https://playwright.dev/docs/api/class-locator#locator-is-visible

so I will leave this up to you to decide:
whenever `isVisible()` is bugged or documentation needs updating.

Through method of deduction my initial context was that because `isVisible` behaviour is NOT in line with documentation, I concluded INCORRECTLY that there is issue with `waitFor` as I was comparing the two as per all of my code examples.
My fault was that I did not pay attention to what documentation states about return value of `waitFor`, it returns `Promise<void>` rather than `Promise<boolean>`, I find it difficult to anticipate behaviour of these methods because of lack of consistency in API.
When I worked on my test, I have found `.isVisible()` in the docs I noticed that it is discouraged due to potential of flakiness, thus I made the conscious choice to use  `waitFor` instead of `isVisible()`.

I hope as you can see by my initial example my goal was to achieve some sort of conditional assertion related to element appearance.

I think that having return value on that method `Promise<void>` rather than `Promise<boolean>` is worse because as I pointed out it complicates conditional testing. I hope you can see my point here? Could that be a update/improvement?

The main argument for `waitFor()` to return `Promise<boolean>` is that it gives us developers more flexibility in handling what happens to particular test.

Sorry but I have also asked a question that you omitted from my previous comment, @dgozman please can you clarify, are you stating that the correct way of using waitFor is without `expect()` assertion?

**Comment by dgozman** [comment]
@renhyl Thank you for the explanation.

> After all the testing I did I have arrived at the conclusion that isVisible() is not working correctly because it awaits timeout but it should not according to documentation

I believe this is not the case. If you can provide a repro that shows that `isVisible()` is waiting, we will look into it. However, we have extensive test coverage that suggests the opposite.

> please can you clarify, are you stating that the correct way of using waitFor is without expect() assertion?

Yes. [Here is a list of assertions](https://playwright.dev/docs/test-assertions#auto-retrying-assertions), but `waitFor()` is just a method on `Locator`.

> I think that having return value on that method Promise<void> rather than Promise<boolean> is worse because as I pointed out it complicates conditional testing.

We will not change the return value of `waitFor()`. There is `isVisible()` that returns `Promise<boolean>` for this usecase. It is discouraged, because conditional testing is usually flaky - the page changes asynchronously so checking whether something is visible may be too early or too late and it's hard to control for that. However, if you cannot do anything else but conditional testing, by all means use `isVisible()` because it's fast and gives you exactly what you need.

---

Given the discussion above, I'll close this issue. There is no action item for Playwright at this time. If you believe that `isVisible()` does not work as stated in the documentation, please fie a new issue by filling in the "Bug Report" template and providing a repro that we can run locally. Thank you for filing this issue!
